### PR TITLE
Add Siamese Network Training example

### PR DESCRIPTION
**Pull request by AminNasiri63**
## Summary
This example demonstrates image similarity estimation using Siamese Networks with triplet loss, featuring distributed training and comprehensive evaluation.

## Key Features
- Triplet loss implementation for similarity learning
- Efficient tf.data pipeline for triplet generation  
- Distributed training across multiple GPUs using MirroredStrategy
- Progressive training approach (frozen backbone → fine-tuning)
- Transfer learning with ResNet50 backbone
- Comprehensive evaluation with confusion matrices and metrics

## Use Cases
- Face recognition and verification
- Person re-identification
- Product similarity matching
- General image similarity learning tasks

## Testing
- [x] Code runs without errors
- [x] Generates expected outputs
- [x] Documentation is clear and comprehensive
- [x] Follows Keras example conventions

## Technical Details
- Uses tf.data for optimized data loading
- Implements custom Keras loss and layers
- Supports multi-GPU training
- Includes visualization and evaluation utilities

This example would be valuable for the Keras community as it demonstrates advanced techniques like triplet loss, distributed training, and comprehensive evaluation in a practical, educational format.

### CODE REVIEW

File: examples/vision/siamese_training.py [hunk 1]
```diff
@@ -0,0 +1,733 @@
...
+import glob
+import random
+import numpy as np
+import matplotlib.pyplot as plt
+import seaborn as sns
+
+import tensorflow as tf
+import tensorflow_datasets as tfds
+from tensorflow.keras.callbacks import ModelCheckpoint
+from keras.saving import register_keras_serializable
```
[the remark this answers is not in the document]
Comment (AminNasiri63) [reply examples/vision/siamese_training.py]: Addressed, thank you!

File: examples/vision/siamese_training.py [hunk 2]
```diff
@@ -0,0 +1,733 @@
...
+"""
+## Dataset Management
+
+The `DatasetManager` class handles dataset directory structure and creates train/validation/test splits.
+"""
+
+class DatasetManager:
+    """Manages dataset directory structure and train/val/test splitting."""
+
+    def __init__(self, dataset_path, min_images=20, split_ratio=(0.7, 0.2, 0.1)):
```
[the remark this answers is not in the document]
Comment (AminNasiri63) [reply examples/vision/siamese_training.py]: Addressed, thank you!

File: examples/vision/siamese_training.py [hunk 3]
```diff
@@ -0,0 +1,733 @@
...
+        super().__init__(**kwargs)
+        self.axis = axis
+
+    def call(self, inputs):
+        return tf.math.l2_normalize(inputs, axis=self.axis)
+
+    def get_config(self):
+        config = super().get_config()
+        config.update({"axis": self.axis})
+        return config
```
[the remark this answers is not in the document]
Comment (AminNasiri63) [reply examples/vision/siamese_training.py]: Used the built-in tf.keras.layers.UnitNormalization layer, thank you!

File: examples/vision/siamese_training.py [hunk 4]
```diff
@@ -0,0 +1,733 @@
...
+    embedding_model = None
+    for layer in model.layers:
+        if hasattr(layer, 'name') and layer.name == 'embedding':
+            embedding_model = layer
+            embedding_model.save("./embedding_model.keras") # Save the trained model to a file named "embedding_model.keras"
+            break
+
+    if not embedding_model:
+        print("Embedding model not found!")
+        return
```
[the remark this answers is not in the document]
Comment (AminNasiri63) [reply examples/vision/siamese_training.py]: Addressed the issue, and the saving step is handled separately. Thank you!

File: examples/vision/siamese_training.py [hunk 5]
```diff
@@ -0,0 +1,733 @@
...
+        print(f"[{mode}] TripletGenerator: {len(self.class_images)} classes")
+
+    def _load_class_images(self):
+        """Load and sample images for each class."""
+        class_images = {}
+
+        for class_folder in self.class_folders:
+            image_files = glob.glob(os.path.join(class_folder, "*.jpg")) + \
+                         glob.glob(os.path.join(class_folder, "*.jpeg")) + \
+                         glob.glob(os.path.join(class_folder, "*.png"))
```
[the remark this answers is not in the document]
Comment (AminNasiri63) [reply examples/vision/siamese_training.py]: Addressed, thank you!

File: examples/vision/siamese_training.py [hunk 6]
```diff
@@ -0,0 +1,733 @@
...
+we need to provide it in the `custom_objects` dictionary
+so Keras knows how to reconstruct that layer when loading.
+"""
+def load_embedding_network(model_path="./embedding_model.keras"):
+    trainedEmbedding = tf.keras.models.load_model(
+        model_path,
+        custom_objects={"L2Normalization": L2Normalization}
+    )
+
+    return trainedEmbedding
```
[the remark this answers is not in the document]
Comment (AminNasiri63) [reply examples/vision/siamese_training.py]: Addressed, thank you!

File: examples/vision/siamese_training.py [hunk 7]
```diff
@@ -0,0 +1,733 @@
...
+            # Compute anchor-negative distances
+            an_distances = tf.reduce_sum(tf.square(indicator - neg_samples), axis=-1)
+
+            # Compare distances - we want ap_distance < an_distance
+            comparisons = tf.stack([ap_distances, an_distances])
+            min_indices = tf.argmin(comparisons, axis=0)
+
+            # Count correct predictions
+            correct_count = tf.reduce_sum(1 - min_indices).numpy()
+            incorrect_count = tf.reduce_sum(min_indices).numpy()
```
[the remark this answers is not in the document]
Comment (AminNasiri63) [reply examples/vision/siamese_training.py]: Addressed, thank you!

File: examples/vision/siamese_training.py [hunk 8]
```diff
@@ -0,0 +1,733 @@
...
+
+# For evaluation, we need test data
+dataset_manager = DatasetManager("./sample_data", Config.MIN_IMAGES_PER_CLASS)
+_, _, test_folders = dataset_manager.get_splits()
+
+if test_folders:
+    print("\n=== Model Evaluation ===")
+    evaluate_model(trained_model, test_folders, Config())
+else:
+    print("No test data available for evaluation")
```
[the remark this answers is not in the document]
Comment (AminNasiri63) [reply examples/vision/siamese_training.py]: Addressed, thank you!

File: examples/vision/siamese_training.py [hunk 9]
```diff
@@ -0,0 +1,733 @@
...
+        return class_images
+
+    def generate_triplets(self):
+        """Generator yielding triplet paths."""
+        class_names = list(self.class_images.keys())
+
+        while True:
+            # Select anchor class and different negative class
+            anchor_class = random.choice(class_names)
+            negative_class = random.choice([c for c in class_names if c != anchor_class])
```
[the remark this answers is not in the document]
Comment (AminNasiri63) [reply examples/vision/siamese_training.py]: Addressed, thank you!

File: examples/vision/siamese_training.py [hunk 10]
```diff
@@ -0,0 +1,733 @@
...
+            tf.TensorSpec(shape=(), dtype=tf.string),
+            tf.TensorSpec(shape=(), dtype=tf.string),
+            tf.TensorSpec(shape=(), dtype=tf.string)
+        )
+    )
+
+    # Apply processing pipeline
+    dataset = dataset.map(processor.process_triplet, num_parallel_calls=tf.data.AUTOTUNE)
+    dataset = dataset.take(config.TRIPLETS_PER_EPOCH)
+    dataset = dataset.cache()
```
[the remark this answers is not in the document]
Comment (AminNasiri63) [reply examples/vision/siamese_training.py]: Addressed, thank you!

**Comment by AminNasiri63** [comment]
Hello Keras Team,

Thank you for the detailed feedback. I’ve addressed all the points and pushed the updated changes to this PR. I really learned a lot from your suggestions during this process.
Could you please review it again when you have a chance? Let me know if any further updates are needed.

Best regards,
Amin

**Comment by AminNasiri63** [comment]
Hi @sachinprasadhs
A heads-up that the updates are ready for review. 

Thanks!
Amin

**Comment by AminNasiri63** [comment]
Hi @sachinprasadhs and @keras-team,

Just a friendly reminder that all checks (including CLA) have passed and the updates are ready for review. I’m happy to adjust anything you need. Please let me know if there’s anything else I should improve on this PR.
Thank you so much for your time!

Amin

**Comment by sachinprasadhs** [comment]
Hi, Thanks for the PR, but we already have this tutorial published in Keras here https://keras.io/examples/vision/siamese_network/
